python/src/equistore/operations/unique_metadata.py

```diff
@@ -0,0 +1,263 @@
+
+    .. code-block:: python
+
+        unique_samples = unique_metadata_block(
+            block, axis="samples", names=["structure"],
+        )
+
+    To find the unique ``"atom"`` indices along the ``"samples"`` axis present
+    in the ``"positions"`` gradient block of a given :py:class:`TensorBlock`:
+
+    .. code-block:: python
+
+        unique_grad_samples = unique_metadata_block(
+            block, axis="samples", names=["atom"], gradient_param="positions",
+        )
+
+    :param block: the :py:class:`TensorBlock` to find unique indices for.
+    :param axis: a str, either ``"samples"`` or ``"properties"``, corresponding
+        to the ``axis`` along which the named unique metadata should be found.
+    :param names: a ``str``, ``list`` of ``str``, or ``tuple`` of ``str``
+        corresponding to the name(s) of the metadata along the specified
+        ``axis`` for which the unique indices should be found.
+    :param gradient_param: a ``str`` corresponding to the gradient parameter
+        name for the gradient blocks to find the unique metadata for. If none
+        (default), the unique metadata of the regular :py:class:`TensorBlock`
+        objects will be calculated.
+
+    :return: a sorted :py:class:`Labels` object containing the unique metadata
+        for the input ``block`` or its gradient for the specified parameter.
+        Each element in the returned :py:class:`Labels` object has
+        len(``names``) entries.
+    """
+    # Parse input args
+    if not isinstance(block, TensorBlock):
+        raise TypeError("`block` must be an equistore `TensorBlock`")
+    names = (
+        [names]
+        if isinstance(names, str)
+        else (list(names) if isinstance(names, tuple) else names)
+    )
+    _check_args(block, axis, names, gradient_param)
+    # Make a list of the blocks to find unique indices for
+    if gradient_param is None:
+        blocks = [block]
+    else:
+        blocks = [block.gradient(gradient_param)]
+
+    return _unique_from_blocks(blocks, axis, names, gradient_param)
+
+
+def _unique_from_blocks(
+    blocks: List[TensorBlock],
+    axis: str,
+    names: List[str],
+    gradient_param: Optional[str],
+) -> Labels:
+    """
+    Finds the unique metadata of a list of blocks along the given ``axis`` and
+    for the specified ``names``. If ``gradient_param`` is specified, only finds
+    the unique indices for gradient blocks under the specified parameter name.
+    """
+    # Extract indices from each block
+    all_idxs = []
+    for block in blocks:
+        if axis == "samples":
+            idxs = block.samples[names]
+        else:  # "properties"
+            idxs = block.properties[names]
+        all_idxs += idxs.tolist()
+
+    # If no matching indices across all blocks return a empty Labels w/ the
+    # correct names
+    if len(all_idxs) == 0:
+        # Create Labels with single entry
+        labels = Labels(names=names, values=np.arange(len(names)).reshape(-1, 1))
```

Comment:
Are we sure that this is correct with `reshape(-1, 1)`. I just tried the code and got and error. 

```python
from equistore import Labels

names = ["1", "l", "f"]
labels = Labels(names=names, values=np.arange(len(names)).reshape(-1, 1))
```

To get a propper Labels instance I had to swap to `reshape(1, -1)`:

```python
names = ["1", "l", "f"]
labels = Labels(names=names, values=np.arange(len(names)).reshape(1, -1))
```

Reply:
Yes, good spot! Something missed by my unit tests so I'll get a test in for that now too :)